### ISSUE TITLE
SOCKS proxy support

### ISSUE THREAD

**Issue by sharkone** [sabnzbd/sabnzbd#206]
I think it would be great to add SOCKS proxy support.

I saw this post which seems to suggest that it should be a simple change:
https://forums.sabnzbd.org/viewtopic.php?f=4&t=3717

Maybe the only thing missing is UI support?

I guess I could look into it if needed.

**Comment by Crafter6432** [comment]
Any reason why this wouldn't be added ?

**Comment by Safihre** [comment]
None of us use it so we can't really test it. And nobody else made a pull request to add it.. 
You're very welcome to do so :)

**Comment by Safihre** [comment]
I will re-open and maybe take it as an optional feature for 2.4.0 / 3.0.0.

**Comment by CoderKiwi** [comment]
Any gauge on how likely this is to be implemented for the upcoming release @Safihre? I think quite a few users would appreciate this feature (me included).

**Comment by Safihre** [comment]
Hmm not sure. How's your python skills? We welcome pull requests if anyone wants to add it!

**Comment by CoderKiwi** [comment]
I actually did edit the source code in `newswrapper.py` as outlined in the now-ancient forum posts that @sharkone referenced. It was pretty straightforward. I got SOCKS5 proxying working through HTTP, but couldn't figure out how to verify it working through SSL. 

But that was hacky. I hardcoded proxy URLs and credentials within the code. So the main issue is how to add the GUI options to provide a SOCKS5 proxy and make those config options be used by `newswrapper.py` instead of hardcoding values. I haven't looked into how to do that. 

Seems like having me (try to) implement it is the least efficient way to do it though. I haven't actually ever submitted a pull request. The SAB devs should be able to just follow the guide at https://forums.sabnzbd.org/viewtopic.php?f=4&t=3717 to program it. 

I read elsewhere that the main obstacle isn't programming it but testing it. If I donate some money specifically for SOCKS5 proxy subscriptions/trials (i.e. TorGuard) would that make this feature more feasible?

**Comment by CoderKiwi** [comment]
I've read elsewhere that the main obstacle to implementing this feature is a proxy to use for testing. I have credentials to both an HTTP(S) and SOCKS5 proxy service which I can share to help implement this. Would you be interested in that @Safihre?

**Comment by CoderKiwi** [comment]
It wouldn't add anything over an existing VPN; that's not really the point. The point of a SOCK5 proxy is to use _instead_ of a VPN. With the encryption of SSL, and the IP address of the proxy, it's essentially as strong as a VPN (in my opinion).

As far as I understand, the use case for a SOCKS5 proxy instead of a VPN has two components. The first is it would be easier to set up for most users. Subscribing to a proxy service and entering credentials into the SABnzbd GUI would be very easy. This is in contrast to using Docker swarm overlay networks to forward traffic through proxy/VPN client containers, or messing with IP tables on the OS of the machine, etc.

The other is speed. SOCKS5 is a lower lever protocol than, say HTTP, NNTP or other protocols, which means, in theory, downloading via SOCKS5 proxy vs a VPN (with all the other conditions being the same) would be faster. In the forums a few people mentioned ~10% speed increases. 

This is just from what I've read/researched.

**Comment by thezoggy** [comment]
Adding full socks support would be easier if we used requests library and if cherrypy had tie ins.. to add it now would be an additional layer of spaghetti code... And with no test sets... This is more than likely going to break stuff than anything. This would better off adding during the code rewrite / python3 move

**Comment by Safihre** [comment]
@thezoggy I don't think it's that bad at all actually.
Look at https://github.com/Anorov/PySocks, it handles both HTTP and SOCKS proxies and it allows you to simply monkey-patch the `socket`.
It's more the question if we want to support proxies in our software directly..

**Comment by CoderKiwi** [comment]
To show it works/is reasonably easy to implement, here's how I made the SOCKS5 proxy work (taken from the [forum](https://forums.sabnzbd.org/viewtopic.php?f=4&t=3717)). I only got it to work via non-SSL; couldn't verify it working using SSL. This is hacky, but it shows it works. I'm sure you'll have a better solution!

First, add an import statement (I don't think this requires PySocks at all):

`import socks` 

Within `newswrapper.py` there is this block within the **NNTP** class to establish a connection :

```
if sslenabled:
    # Use context or just wrapper
    if sabnzbd.HAVE_SSL_CONTEXT:
        # Setup the SSL socket
        ctx = ssl.create_default_context()

        # Only verify hostname when we're strict
        if(nw.server.ssl_verify < 2):
            ctx.check_hostname = False
        # Certificates optional
        if(nw.server.ssl_verify == 0):
            ctx.verify_mode = ssl.CERT_NONE

        # Did the user set a custom cipher-string?
        if(nw.server.ssl_ciphers):
            # At their own risk, socket will error out in case it was invalid
            ctx.set_ciphers(nw.server.ssl_ciphers)

        self.sock = ctx.wrap_socket(socket.socket(af, socktype, proto), server_hostname=str(nw.server.host))
    else:
        # Ciphers have to be None, if set to empty-string it will fail on <2.7.9
        ciphers = nw.server.ssl_ciphers if nw.server.ssl_ciphers else None
        # Use a regular wrapper, no certificate validation
        self.sock = ssl.wrap_socket(socket.socket(af, socktype, proto), ciphers=ciphers)
else:
    self.sock = socket.socket(af, socktype, proto)
```

Simply change the very last line to use a SOCKS5 socket instead, replacing the fields in <> with the SOCKS5 proxy credentials:

```
else:
    self.sock = socks.socksocket(af, socktype, proto)     
    self.sock.setproxy(socks.PROXY_TYPE_SOCKS5,"<insert_domain_name>",<port>,True,"<username>","<password")
```

With that changed, connections to a usenet server, when SSL is unchecked, go through the proxy.

**Comment by Safihre** [comment]
Why doesn't ssl work?

**Comment by CoderKiwi** [comment]
It's not that it doesn't work through SSL, I just couldn't verify it working. For some reason -- even with bad credentials to the proxy -- it connected to the internet somehow. So I began thinking I just didn't understand the rest of the code and was implementing it incorrectly. 

This has to do with lines **189-194** (the if-else blocks with SSL context and without).

The trick with SSL is to set the proxy and everything before establishing the SSL connection. Definitely cannot set up a proxy after the SSL connection is established. 

At this point I've told you basically everything I know. Hopefully you have some more insight into this.

**Comment by thezoggy** [comment]
@Safihre I'm talking about the fact that we do not have one url handler across all facets of the app. Adding proxy/socks support to the newswrapper is just one part..  what about rss reader.. what about login auth.. what about checking for updates.. etc etc etc. can see where if we add it in one place someone is going to complain x doesnt work else where.

https://github.com/sabnzbd/sabnzbd/search?q=urllib&unscoped_q=urllib

**Comment by CoderKiwi** [comment]
Any movement on this?

**Comment by Safihre** [comment]
@CoderKiwi I think @thezoggy has a point that it indeed should work for all sockets. But I'm not sure if the library supports all types (IPv6, SSL)?

**Comment by Aidolii** [comment]
Still nothing? :(

**Comment by christian-weiss** [comment]
Sabnzbd should follow "secure by default" pattern.

SOCKS5 is less secure then VPN.
Read https://thevpn.guru/socks5-proxy-vs-vpn-difference/ for a nice explanation.

If you can not use your VPN service because of its IP ranges are not accepted by your newsgroup service then switch the newsgroup service or switch to anoter VPN service. 

I put my vote for closing this ticket. Limited sabnzbd team capacity should work on more important stuff then on drilling security holes.

**Comment by ImportTaste** [comment]
Is this still in the cards?

**Comment by MetalSpork** [comment]
The lack of proxy support means putting the whole system behind a VPN-linked network interface, which, while great for privacy, is certainly not desirable when latency needs to be kept to a minimum. SOCKS5, assuming you're using custom DNS, is secure enough for usenet downloading. But it's a pain that most of the major nzb download tools do not support it.

Please exhume this from the backlog, SABnzbd is my favorite among its competitors, but without SOCKS5 proxy support, I have to use something crappy instead.

**Comment by sjpotter** [comment]
I've hacked up the code per the instructions above, and it does work, but its static (i.e. my server/username/password) shouldn't be hard to turn into variables and hook into the UI, but I don't know how to do that.

though the first cut of it would be socks in replacement of ssl (i.e. incompatible with each other), as the ssl code is making the connection itself right now, not operating on an existing socket.

**Comment by ImportTaste** [comment]
> I've hacked up the code per the instructions above, and it does work, but its static (i.e. my server/username/password) shouldn't be hard to turn into variables and hook into the UI, but I don't know how to do that.
> 
> though the first cut of it would be socks in replacement of ssl (i.e. incompatible with each other), as the ssl code is making the connection itself right now, not operating on an existing socket.

I'd be happy with just environmental variables at this point.

**Comment by dimqua** [comment]
It would be nice to have a possibility to use a proxy only for getting nzb files from indexers. I use an indexer that is blocked in my country and I can't access it directly.

**Comment by slurikadre** [comment]
An example of where this could be really handy is when used in conjunction with https://github.com/wfg/docker-openvpn-client

**Comment by puzzledsab** [comment]
I hope you'll help test my SOCKS 5 pull request, see https://github.com/sabnzbd/sabnzbd/pull/1894 for info.

**Comment by instylegithub** [comment]
Hi,  I would really like this feature too.  This option would be a lot easier to setup VS a split tunnel VPN.

**Comment by Safihre** [comment]
Anyone willing to test #1961?

**Comment by Safihre** [comment]
Implemented as part of 3.5.0.
